src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/CustomL2H.java

```diff
@@ -0,0 +1,2062 @@
+        this.docLabelCounts = new SparseCount[D];
+        this.docMaskes = new Set[D];
+
+        for (int d = 0; d < D; d++) {
+            this.z[d] = new int[words[d].length];
+            this.x[d] = new int[words[d].length];
+            this.docSwitches[d] = new DirMult(
+                    new double[]{hyperparams.get(A_0), hyperparams.get(B_0)});
+            this.docLabelCounts[d] = new SparseCount();
+            this.docMaskes[d] = new HashSet<Integer>();
+
+            // Hui Chen: Set<Integer> cands = getCandidates(initPredictions[d], topK); -> 2 statements
+            // int actualK = topK <= initPredictions[d].length ? topK : initPredictions[d].length;
+            // Set<Integer> cands = getCandidates(initPredictions[d], actualK);
+            Set<Integer> cands = getCandidates(initPredictions[d], topK);
+            for (int label : cands) {
+                Node node = nodes[label];
+                while (node != null) {
+                    docMaskes[d].add(node.id);
+                    node = node.getParent();
+                }
+            }
+        }
+
+        // initialize: sampling using global distribution over labels
+        if (verbose) {
+            logln("--- Sampling on test data ...");
+        }
+        double[][] predictedScores = new double[D][L - 1]; // exclude root
+        int count = 0;
+        for (iter = 0; iter < testMaxIter; iter++) {
+            if (iter % testSampleLag == 0) {
+                logln("--- --- iter " + iter + "/" + testMaxIter
+                        + " @ thread " + Thread.currentThread().getId()
+                        + "\t" + getSamplerFolderPath());
+            }
+            if (iter == 0) {
+                sampleXZsMH(!REMOVE, !ADD, !REMOVE, ADD);
+            } else {
+                sampleXZsMH(!REMOVE, !ADD, REMOVE, ADD);
+            }
+
+            if (iter >= this.testBurnIn && iter % this.testSampleLag == 0) {
+                for (int dd = 0; dd < D; dd++) {
+                    for (int ll = 0; ll < L - 1; ll++) {
+                        predictedScores[dd][ll]
+                                += (double) docLabelCounts[dd].getCount(ll) / words[dd].length;
+                    }
+                }
+                count++;
+            }
+        }
+
+        // output result during test time
+        if (verbose &&  outputResultFile != null) {
+            logln("--- Outputing result to " + outputResultFile);
+        }
+        for (int dd = 0; dd < D; dd++) {
+            for (int ll = 0; ll < L - 1; ll++) {
+                predictedScores[dd][ll] /= count;
+            }
+        }
+		if(outputResultFile != null) {
+			PredictionUtils.outputSingleModelClassifications(new File(outputResultFile),
+					predictedScores);
+			return null;
+		} else {
+			return outputSingleModelClassificationsToString(predictedScores);
+		}
+    }
+
+	private static String outputSingleModelClassificationsToString(double[][] predictions) {
```

Comment:
Serializing a 2d array to a string to be later deserialized seems a bit inefficient. Could you not just return the 2d array directly and skip the string serialization?